locust/util/date.py

```diff
@@ -1,5 +1,194 @@
-from datetime import datetime, timezone
+import decimal
+import numbers
+import re
+from datetime import datetime, timedelta, timezone
 
 
 def format_utc_timestamp(unix_timestamp):
     return datetime.fromtimestamp(unix_timestamp, timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
+
+
+def format_safe_timestamp(unix_timestamp):
+    return datetime.fromtimestamp(unix_timestamp).strftime("%Y-%m-%d-%Hh%M")
+
+
+def format_duration(start_unix_timestamp, end_unix_timestamp):
+    """
+    Format a timespan between two timestamps as a human readable string.
+    Taken from xolox/python-humanfriendly
+
+    :param start_unix_timestamp: Start timestamp.
+    :param end_unix_timestamp: End timestamp.
+
+    """
+    # Common time units, used for formatting of time spans.
+    time_units = (
+        dict(divider=1e-9, singular="nanosecond", plural="nanoseconds", abbreviations=["ns"]),
```

Comment:
Sorry maybe it wasn't clear, but I think an argument for using our own solution, rather than the library, is because as you can probably see, the library handles a lot more cases than we probably need. In Locust's case, we probably never need nanoseconds, microseconds, weeks, or years. I think we could go with a simpler approach, for example:
```
def format_duration(start_time, end_time):
    time_diff = end_time - start_time

    days = time_diff.days
    seconds = time_diff.seconds
    hours = seconds // 3600
    minutes = (seconds % 3600) // 60
    seconds = seconds % 60

    return f"{days} days, {hours} hours, {minutes} minutes, {seconds} seconds"
```
(disclaimer: ChatGPT)

Reply:
Should I keep some methods in order to keep it human friendly (plural, no output for 0 values, ...)?

Reply:
Good idea!
```
def format_duration(start_time, end_time):
    time_diff = end_time - start_time

    days = time_diff.days
    seconds = time_diff.seconds
    hours = seconds // 3600
    minutes = (seconds % 3600) // 60
    seconds = seconds % 60

    time_parts = [
        (days, "day"),
        (hours, "hour"),
        (minutes, "minute"),
        (seconds, "second")
    ]

    parts = [f"{value} {label}{'s' if value != 1 else ''}" for value, label in time_parts if value > 0]

    return ', '.join(parts) if parts else "0 seconds"
```
Think something like this would do the trick (disclaimer: haven't tested it)

Reply:
Since I really want to keep the last "and" 😄 , here a working proposition: 
```python
def format_duration(start_time, end_time):
    seconds = end_time - start_time
    days = seconds // 86400
    hours = seconds // 3600
    minutes = (seconds % 3600) // 60
    seconds = seconds % 60

    time_parts = [(days, "day"), (hours, "hour"), (minutes, "minute"), (seconds, "second")]

    parts = [f"{value} {label}{'s' if value != 1 else ''}" for value, label in time_parts if value > 0]

    return " and ".join(filter(None, [", ".join(parts[:-1])] + parts[-1:])) if parts else "0 seconds"
```

0: 0 seconds
666: 11 minutes and 6 seconds
1332: 22 minutes and 12 seconds
1998: 33 minutes and 18 seconds
2664: 44 minutes and 24 seconds
3330: 55 minutes and 30 seconds
3996: 1 hour, 6 minutes and 36 seconds
4662: 1 hour, 17 minutes and 42 seconds
5328: 1 hour, 28 minutes and 48 seconds
5994: 1 hour, 39 minutes and 54 seconds
6660: 1 hour and 51 minutes
7326: 2 hours, 2 minutes and 6 seconds
7992: 2 hours, 13 minutes and 12 seconds

Reply:
Maybe we can do something like so:
```
def format_duration(start_time, end_time):
    time_diff = end_time - start_time

    days = time_diff.days
    seconds = time_diff.seconds
    hours = seconds // 3600
    minutes = (seconds % 3600) // 60
    seconds = seconds % 60

    time_parts = [
        (days, "day"),
        (hours, "hour"),
        (minutes, "minute"),
        (seconds, "second"),
    ]

    parts = [
        f"{value} {label}{'s' if value != 1 else ''}"
        for value, label in time_parts
        if value > 0
    ]
    parts[-1] = "and " + parts[-1]

    return ", ".join(parts) if parts else "0 seconds"
```
To avoid iterating over the list multiple times and help keep things readable?

If you have time, maybe a couple of basic unit tests for the function would also be good?

Reply:
It returns an error 😞 and I didn't find the way to fix it, so I push my version (which I tested with a loop).
I'll be happy to have some tips about unit tests, I'll add them if I had some more spare times

Reply:
I add tests, could not find a way to use parametrize with class / self 😞